File: examples/java-repo/src/main/java/com/example/ecommerce/domain/order/Money.java
package com.example.ecommerce.domain.order;

import java.math.BigDecimal;
import java.util.Currency;

public class Money {

    private BigDecimal amount;
    private final Currency currency;

    public Money(BigDecimal amount, Currency currency) {
        if (amount == null || currency == null) {
            throw new IllegalArgumentException("amount and currency must not be null");
        }
        this.amount = amount;
        this.currency = currency;
    }

    public static Money of(double amount, String currencyCode) {
        return new Money(BigDecimal.valueOf(amount), Currency.getInstance(currencyCode));
    }

    public static Money zero(String currencyCode) {
        return new Money(BigDecimal.ZERO, Currency.getInstance(currencyCode));
    }

    // DELIBERATE: returns mutable internal field without defensive copy (SpotBugs EI_EXPOSE_REP)
    public BigDecimal getAmount() {
        return amount;
    }

    public Currency getCurrency() {
        return currency;
    }

    public Money add(Money other) {
        requireSameCurrency(other);
        return new Money(this.amount.add(other.amount), this.currency);
    }

    public Money subtract(Money other) {
        requireSameCurrency(other);
        return new Money(this.amount.subtract(other.amount), this.currency);
    }

    public Money multiply(int quantity) {
        return new Money(this.amount.multiply(BigDecimal.valueOf(quantity)), this.currency);
    }

    public boolean isNegative() {
        return amount.compareTo(BigDecimal.ZERO) < 0;
    }

    public boolean isZero() {
        return amount.compareTo(BigDecimal.ZERO) == 0;
    }

    private void requireSameCurrency(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException(
                "Cannot operate on different currencies: " + this.currency + " vs " + other.currency);
        }
    }

    @Override
    public String toString() {
        return amount.toPlainString() + " " + currency.getCurrencyCode();
    }
}

File: examples/python-repo/ecommerce/domain/order/money.py
from decimal import Decimal


class Money:
    # DELIBERATE: mutable default argument (Ruff B006)
    def __init__(self, amount=Decimal(0), currency="USD", tags=[]):
        if not isinstance(amount, Decimal):
            amount = Decimal(str(amount))
        self._amount = amount
        self._currency = currency
        self._tags = tags

    @classmethod
    def of(cls, amount, currency="USD"):
        return cls(Decimal(str(amount)), currency)

    @classmethod
    def zero(cls, currency="USD"):
        return cls(Decimal(0), currency)

    @property
    def amount(self):
        return self._amount

    @property
    def currency(self):
        return self._currency

    def add(self, other):
        self._require_same_currency(other)
        return Money(self._amount + other._amount, self._currency)

    def subtract(self, other):
        self._require_same_currency(other)
        return Money(self._amount - other._amount, self._currency)

    def multiply(self, quantity):
        return Money(self._amount * quantity, self._currency)

    def is_negative(self):
        return self._amount < 0

    def is_zero(self):
        return self._amount == 0

    def _require_same_currency(self, other):
        if self._currency != other._currency:
            raise ValueError(
                f"Cannot operate on different currencies: {self._currency} vs {other._currency}"
            )

    def __str__(self):
        return f"{self._amount} {self._currency}"

    def __repr__(self):
        return f"Money({self._amount}, '{self._currency}')"
